### PR TITLE
fix(facets): implement the possibility to set a strategy for what filters are sent to the request. leaves-only strategy also implemented

### PR DESCRIPTION
**Pull request by CachedaCodes**
EX-7550

## Motivation and context
Sometimes the backend doesn't handle the hierarchical filters sent as expected and only filters by the parent. This PR implements the posibility of applying a strategy to filter the filters that are sent to the requests or that are stored in the URL 

- [ ] Dependencies. If any, specify:
- [X] Open issue. If applicable, link:

## Type of change

- [X] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Change requires a documentation update

## What is the destination branch of this PR?

- [X] `Main`
- [ ] Other. Specify:

## How has this been tested?
The strategy is set to `leaves-only`, which means that only filters without selected children are sent to the request or stored in the URL. Check the request or the url while selecting filters and change the strategy to `all` in the `base-config.ts`.

### CODE REVIEW

File: packages/x-components/src/x-modules/facets/store/module.ts [hunk 1]
```diff
@@ -17,10 +18,14 @@ export const facetsXStoreModule: FacetsXStoreModule = {
     groups: {},
     facets: {},
     preselectedFilters: [],
-    stickyFilters: {}
+    stickyFilters: {},
+    config: {
+      filtersForRequestStrategy: 'all'
```
Comment (diegopf) [inline packages/x-components/src/x-modules/facets/store/module.ts:23]: Shouldn't we default to the `leaves-only`? 🤔

File: packages/x-components/src/x-modules/facets/store/getters/selected-filters-for-request.getter.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Default implementation for the {@link FacetsGetters.selectedFiltersForRequest} getter.
+ *
+ * @param state - Current {@link https://vuex.vuejs.org/guide/state.html | state} of the facets'
+ * module.
+ *
+ * @returns A list containing the selected filters that conform to the filters for request strategy.
+ *
+ * @public
+ */
+export const selectedFiltersForRequest: FacetsXStoreModule['getters']['selectedFilters'] = state =>
```
Comment (diegopf) [inline packages/x-components/src/x-modules/facets/store/getters/selected-filters-for-request.getter.ts:14]:
```suggestion
export const selectedFiltersForRequest: FacetsXStoreModule['getters']['selectedFiltersForRequest'] = state =>
```

File: packages/x-components/src/x-modules/facets/store/getters/selected-filters-for-request.getter.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        state.config.filtersForRequestStrategy === 'leaves-only' &&
+        isHierarchicalFilter(filter)
+      ) {
+        const childrenIds = filter.children?.map(child => child.id);
+
+        return !filters.some(newFilter => childrenIds?.includes(newFilter.id));
+      }
+
+      return true;
+    });
```
Comment (diegopf) [inline packages/x-components/src/x-modules/facets/store/getters/selected-filters-for-request.getter.ts:28]: If we have configured the leaves-only strategy, wouldn't we be performing an unnecessary second filter loop? Perhaps we could consider doing something like

```ts
const selectedFilters = Object.values(state.filters).filter(filter => filter.selected);
    if (state.config.filtersForRequestStrategy === 'leaves-only') {
      return selectedFilters.filter((filter, _, filters) => {
        if (isHierarchicalFilter(filter)) {
          const childrenIds = filter.children?.map(child => child.id);

          return !filters.some(newFilter => childrenIds?.includes(newFilter.id));
        }

        return true;
      });
    }
    return selectedFilters;

**Review by diegopf — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/facets/store/module.ts (hunk 1 above)
Comment (CachedaCodes) [reply packages/x-components/src/x-modules/facets/store/module.ts:23]: Default behavior for the library is `all` and for our implementation is `leaves-only`.

**Review by diegopf — APPROVED** [review]

File: packages/x-components/src/x-modules/facets/store/types.ts [hunk 4]
```diff
@@ -195,3 +207,20 @@ export interface MutateFilterPayload {
...
 }
+
+/**
+ * Configuration options for the {@link FacetsXModule}.
+ *
+ * @public
+ */
+export interface FacetsConfig {
+  /** The filter strategy to use when providing the selected filters for requests. */
+  filtersForRequestStrategy: FiltersForRequestStrategy;
```
Comment (tiborux) [inline packages/x-components/src/x-modules/facets/store/types.ts:218]: should be `filtersStrategyForRequest`? or `filtersStrategy`? or `filtersStrategyRequest`?

Comment (CachedaCodes) [reply packages/x-components/src/x-modules/facets/store/types.ts:218]: I like `filtersStrategyForRequest` better, changing it 👍

**Review by tiborux — APPROVED** [review]

**Review by tiborux — APPROVED** [review]

**Pull request merged**